### PR TITLE
configure_file: Support taking values from a dict

### LINKED ISSUE #4218: RFC: support a dictionary as the configuration parameter of configure_file?
I like dictionaries. Any support for the following idea?
```
configure_file(output: 'config.h', configuration: {'version': '1.0'})
```

Or as a more cluttered alternative (but still better than was is available right now): give the `configuration_data()` constructor an optional dictionary argument:
```
configure_file(output: 'config.h', configuration: configuration_data({'version': '1.0'}))
```

### PR DESCRIPTION
**Pull request by xclaesse**
Closes #4218

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -271,6 +271,10 @@ def __init__(self, pv):
                              'get_unquoted': self.get_unquoted_method,
                              'merge_from': self.merge_from_method,
                              })
+        for name, val in initial_values.items():
+            if not isinstance(val, (int, str)):
+                raise InterpreterException('configuration values must be bool, integer or string')
```
Comment (jpakkane) [inline mesonbuild/interpreter.py]: With this there are now two different locations where the values are checked. This means they can get out of sync. There really should be only place for verifications.

Could this be changed to work roughly like the following:

 - build empty configuration data object
 - if a dict is given, then call the object's `set_method` for each entry in the dict
 - proceed as usual

**Comment by jpakkane** [comment]
Also needs documentation updates in the reference manual, `Configuration.md` as well as a release note snippet.

File: mesonbuild/interpreter.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/interpreter.py]: Good point, PR updated.

**Comment by xclaesse** [comment]
PR updated.

**Comment by jpakkane** [comment]
Reference manual updates are still missing.

**Comment by xclaesse** [comment]
> Reference manual updates are still missing.

Forgot to `git add` that file. PR updated.

**Pull request merged**